tools/unitctl/unitctl/src/cmd/save.rs

```diff
@@ -0,0 +1,52 @@
+use crate::unitctl::UnitCtl;
+use crate::wait;
+use crate::UnitctlError;
+use crate::requests::send_empty_body_deserialize_response;
+use unit_client_rs::unit_client::UnitClient;
+use tar::{Builder, Header};
+use std::fs::File;
+use std::io::stdout;
+
+
+pub async fn cmd(
+    cli: &UnitCtl,
+    filename: &String
+) -> Result<(), UnitctlError> {
+    if !filename.ends_with(".tar") {
+        eprintln!("Warning: writing tarball to {}", filename);
+    }
+
+    let control_socket = wait::wait_for_socket(cli).await?;
+    let client = UnitClient::new(control_socket);
+
+    let config_res = serde_json::to_string(
+        &send_empty_body_deserialize_response(&client, "GET", "/config").await?
```

Comment:
What's the reason for minifying the `export` content?

The Unit control socket applies formatting/prettification to the configuration which is significantly less readable when `export`ed.

```
$ curl --unix-socket /opt/homebrew/var/run/unit/control.sock _/config
{
	"routes": [
		{
			"action": {
				"share": "/Users/l.crilly/containers/unit/fallback/www$uri",
				"fallback": {
					"return": 400
				}
			}
		}
	],

	"listeners": {
		"*:9000": {
			"pass": "routes"
		}
	}
}
$ unitctl export -f - 2> /dev/null | tar xfO - config.json
{"routes":[{"action":{"share":"/Users/l.crilly/containers/unit/fallback/www$uri","fallback":{"return":400}}}],"listeners":{"*:9000":{"pass":"routes"}}}%
```

Why do we need to use serde here at all? Surely we can rely on the control API to return correctly formatted JSON?

Reply:
The CLI can return output in any number formats. I simply use this switch for colored output at my terminal. the output_format feature is not in the scope of this PR. Were you worried that the control API was not outputting proper JSON?

Reply:
Marking this conversation as resolved because the serialization and de-serialization features are not in scope of this PR.

Reply:
So it can display the output as it received it with no processing?

I think your reply is slightly dismissive...

>  Were you worried that the control API was not outputting proper JSON?

I don't believe so, in fact quite the opposite.

The control API is returning properly formatted JSON which was them minified and made much harder to read by unitctl...

So AIUI, the question is. Why if the control API is giving us properly formatted JSON, is unitctl bothering to process it all? (rather than leave any further processing to the likes of jq(1))

Reply:
Unresolving and throwing my hat in with the above.

Export is meant for archiving and exchanging Unit's state; pretty-printing makes manual inspection and editing significantly easier. Since the `/config` endpoint pretty prints itself, not having that in an export feels like a regression versus directly calling curl.

I see two solutions, both equally great:

1. Skip serde and write the raw body bytes from `GET /config`
2. Pretty-print when deserializing.

Reply:
If you just want it pretty printed I am happy to do that. The output of the control api is deserialized which is why the formatting of the json is reset. (and why serde_json is used here).

It'll be significantly more efficient code wise to just re-add pretty printing than to implement custom request code just to have an option that doesnt deserialize the body.